### PR TITLE
Jacobi coordinates

### PR DESCRIPTION
**Pull request by eggrobin**
And `weight` in `BarycentreCalculator`.

### CODE REVIEW

File: geometry/barycentre_calculator.hpp [hunk 1]
```diff
@@ -17,6 +17,8 @@ class BarycentreCalculator {
   void Add(Vector const& vector, Scalar const& weight);
   Vector Get() const;
 
+  Scalar const& weight() const;
```
Comment (pleroy) [inline geometry/barycentre_calculator.hpp]: Document that this is the sum of the weight added so far.

File: physics/degrees_of_freedom_body.hpp [hunk 2]
```diff
@@ -118,6 +118,12 @@ BarycentreCalculator<physics::DegreesOfFreedom<Frame>, Weight>::Get() const {
 }
 
 template<typename Frame, typename Weight>
+inline Weight const&
```
Comment (pleroy) [inline physics/degrees_of_freedom_body.hpp]: inline?

File: physics/degrees_of_freedom_body.hpp [hunk 3]
```diff
@@ -132,5 +138,12 @@ BarycentreCalculator<physics::RelativeDegreesOfFreedom<Frame>, Weight>::Get()
   return implementation_.Get();
 }
 
+template<typename Frame, typename Weight>
+inline Weight const&
```
Comment (pleroy) [inline physics/degrees_of_freedom_body.hpp]: Same.

File: physics/jacobi_coordinates.hpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+// in Jacobi coordinates to barycentric coordinates.
+template<typename Frame>
+class JacobiCoordinates {
+ public:
+  JacobiCoordinates(MassiveBody const& primary);
+
+  // Adds |body| with the given |DegreesOfFreedom| with respect to the
+  // barycentre of the existing bodies.
+  void Add(MassiveBody const& body,
+           RelativeDegreesOfFreedom<Frame> const& dof_wrt_system);
```
Comment (pleroy) [inline physics/jacobi_coordinates.hpp]: wrt -> relative_to

File: physics/jacobi_coordinates.hpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+  // Adds |body| with the |RelativeDegreesOfFreedom| of a |KeplerOrbit| with the
+  // given |KeplerianElements| around the barycentre of the existing bodies.
+  // |osculating_elements_wrt_system| must be a valid argument to the
+  // constructor of |KeplerOrbit|.
+  // Equivalent to
```
Comment (pleroy) [inline physics/jacobi_coordinates.hpp]: I'm not sure that the equivalence helps.  I'd remove it.

File: physics/jacobi_coordinates.hpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      MassiveBody const& body,
+      KeplerianElements<Frame> const& osculating_elements_wrt_system);
+
+  // A body with the total mass of the existing bodies.
+  MassiveBody System() const;
+
+  // Returns the degrees of freedom of the bodies with respect to their
+  // barycentre, in the order in which they were added (starting with the
+  // primary).
+  std::vector<RelativeDegreesOfFreedom<Frame>> BarycentricCoordinates() const;
```
Comment (pleroy) [inline physics/jacobi_coordinates.hpp]: Coordinates is already overloaded.  BarycentricDegreesOfFreedom.

File: physics/jacobi_coordinates_body.hpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+#include "physics/jacobi_coordinates.hpp"
+
+#include <vector>
+
+namespace principia {
+namespace physics {
+
+template<typename Frame>
+JacobiCoordinates<Frame>::JacobiCoordinates(MassiveBody const& primary) {
+  DegreesOfFreedom<PrimocentricFrame> motionless_origin = {
```
Comment (pleroy) [inline physics/jacobi_coordinates_body.hpp]: const

File: geometry/barycentre_calculator.hpp (hunk 1 above)
Comment (eggrobin) [inline geometry/barycentre_calculator.hpp]: Done.

File: physics/degrees_of_freedom_body.hpp (hunk 2 above)
Comment (eggrobin) [inline physics/degrees_of_freedom_body.hpp]: Done.

File: physics/degrees_of_freedom_body.hpp (hunk 3 above)
Comment (eggrobin) [inline physics/degrees_of_freedom_body.hpp]: Done.

File: physics/jacobi_coordinates.hpp (hunk 4 above)
Comment (eggrobin) [inline physics/jacobi_coordinates.hpp]: Done.

File: physics/jacobi_coordinates.hpp (hunk 5 above)
Comment (eggrobin) [inline physics/jacobi_coordinates.hpp]: Done.

File: physics/jacobi_coordinates_body.hpp (hunk 7 above)
Comment (eggrobin) [inline physics/jacobi_coordinates_body.hpp]: Done.

File: physics/jacobi_coordinates.hpp (hunk 6 above)
Comment (eggrobin) [inline physics/jacobi_coordinates.hpp]: Done.

**Pull request merged**